File: backend/schemas/market_overview_trends.py
"""
Pydantic schemas for Market Overview Trends endpoint.
"""

from pydantic import BaseModel


class TrendDataPoint(BaseModel):
    """A single (period, value) data point for a trend chart."""

    period: str
    value: float | None = None


class MarketOverviewTrendsResponse(BaseModel):
    """Pre-aggregated time-series for 27 market trend metrics."""

    # --- Section 1: Financial Trends ---
    companies_listed: list[TrendDataPoint]
    total_market_cap: list[TrendDataPoint]
    total_operating_profit: list[TrendDataPoint]
    total_sales: list[TrendDataPoint]
    total_ebitda: list[TrendDataPoint]
    total_pat: list[TrendDataPoint]
    total_debt: list[TrendDataPoint]
    median_debt_to_equity: list[TrendDataPoint]
    total_net_fixed_assets: list[TrendDataPoint]

    # --- Section 2: Holdings & Returns ---
    median_promoter_holdings: list[TrendDataPoint]
    median_institutional_holdings: list[TrendDataPoint]
    median_public_holdings: list[TrendDataPoint]
    median_ebitda_margin: list[TrendDataPoint]
    median_operating_profit_margin: list[TrendDataPoint]
    median_pat_margin: list[TrendDataPoint]
    median_roe: list[TrendDataPoint]
    median_roce: list[TrendDataPoint]
    median_roa: list[TrendDataPoint]

    # --- Section 3: Cash Flow & Valuation ---
    median_receivable_days: list[TrendDataPoint]
    median_cfo_to_ebitda: list[TrendDataPoint]
    median_cfo_to_pbt: list[TrendDataPoint]
    median_pe: list[TrendDataPoint]
    median_price_to_book: list[TrendDataPoint]
    median_price_to_sales: list[TrendDataPoint]
    median_ev_ebitda: list[TrendDataPoint]
    median_ev_cfo: list[TrendDataPoint]
    median_mcap_netblock: list[TrendDataPoint]

File: frontend-next/src/types/marketOverviewTrends.ts
export interface TrendDataPoint {
  period: string;
  value: number | null;
}

export interface MarketOverviewTrendsResponse {
  // Section 1: Financial Trends
  companies_listed: TrendDataPoint[];
  total_market_cap: TrendDataPoint[];
  total_operating_profit: TrendDataPoint[];
  total_sales: TrendDataPoint[];
  total_ebitda: TrendDataPoint[];
  total_pat: TrendDataPoint[];
  total_debt: TrendDataPoint[];
  median_debt_to_equity: TrendDataPoint[];
  total_net_fixed_assets: TrendDataPoint[];

  // Section 2: Holdings & Returns
  median_promoter_holdings: TrendDataPoint[];
  median_institutional_holdings: TrendDataPoint[];
  median_public_holdings: TrendDataPoint[];
  median_ebitda_margin: TrendDataPoint[];
  median_operating_profit_margin: TrendDataPoint[];
  median_pat_margin: TrendDataPoint[];
  median_roe: TrendDataPoint[];
  median_roce: TrendDataPoint[];
  median_roa: TrendDataPoint[];

  // Section 3: Cash Flow & Valuation
  median_receivable_days: TrendDataPoint[];
  median_cfo_to_ebitda: TrendDataPoint[];
  median_cfo_to_pbt: TrendDataPoint[];
  median_pe: TrendDataPoint[];
  median_price_to_book: TrendDataPoint[];
  median_price_to_sales: TrendDataPoint[];
  median_ev_ebitda: TrendDataPoint[];
  median_ev_cfo: TrendDataPoint[];
  median_mcap_netblock: TrendDataPoint[];
}
